File: python/atlassian_graphql/canonical_models.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional


@dataclass(frozen=True)
class JiraUser:
    account_id: str
    display_name: str
    email: Optional[str] = None


@dataclass(frozen=True)
class JiraProject:
    cloud_id: str
    key: str
    name: str
    type: Optional[str] = None


@dataclass(frozen=True)
class JiraSprint:
    id: str
    name: str
    state: str
    start_at: Optional[str] = None
    end_at: Optional[str] = None
    complete_at: Optional[str] = None


@dataclass(frozen=True)
class JiraIssue:
    cloud_id: str
    key: str
    project_key: str
    issue_type: str
    status: str
    created_at: str
    updated_at: str
    resolved_at: Optional[str] = None
    assignee: Optional[JiraUser] = None
    reporter: Optional[JiraUser] = None
    labels: List[str] = field(default_factory=list)
    components: List[str] = field(default_factory=list)
    story_points: Optional[float] = None
    sprint_ids: List[str] = field(default_factory=list)


@dataclass(frozen=True)
class JiraChangelogItem:
    field: str
    from_value: Optional[str] = None
    to_value: Optional[str] = None
    from_string: Optional[str] = None
    to_string: Optional[str] = None


@dataclass(frozen=True)
class JiraChangelogEvent:
    issue_key: str
    event_id: str
    created_at: str
    items: List[JiraChangelogItem]
    author: Optional[JiraUser] = None


@dataclass(frozen=True)
class JiraWorklog:
    issue_key: str
    worklog_id: str
    started_at: str
    time_spent_seconds: int
    created_at: str
    updated_at: str
    author: Optional[JiraUser] = None


@dataclass(frozen=True)
class OpsgenieTeamRef:
    id: str
    name: str


@dataclass(frozen=True)
class CanonicalProjectWithOpsgenieTeams:
    project: JiraProject
    opsgenie_teams: List[OpsgenieTeamRef] = field(default_factory=list)

File: go/atlassian/canonical_models.go
package atlassian

type JiraUser struct {
	AccountID   string  `json:"accountId"`
	DisplayName string  `json:"displayName"`
	Email       *string `json:"email,omitempty"`
}

type JiraProject struct {
	CloudID string  `json:"cloudId"`
	Key     string  `json:"key"`
	Name    string  `json:"name"`
	Type    *string `json:"type,omitempty"`
}

type JiraSprint struct {
	ID         string  `json:"id"`
	Name       string  `json:"name"`
	State      string  `json:"state"`
	StartAt    *string `json:"startAt,omitempty"`
	EndAt      *string `json:"endAt,omitempty"`
	CompleteAt *string `json:"completeAt,omitempty"`
}

type JiraIssue struct {
	CloudID     string    `json:"cloudId"`
	Key         string    `json:"key"`
	ProjectKey  string    `json:"projectKey"`
	IssueType   string    `json:"issueType"`
	Status      string    `json:"status"`
	CreatedAt   string    `json:"createdAt"`
	UpdatedAt   string    `json:"updatedAt"`
	ResolvedAt  *string   `json:"resolvedAt,omitempty"`
	Assignee    *JiraUser `json:"assignee,omitempty"`
	Reporter    *JiraUser `json:"reporter,omitempty"`
	Labels      []string  `json:"labels"`
	Components  []string  `json:"components"`
	StoryPoints *float64  `json:"storyPoints,omitempty"`
	SprintIDs   []string  `json:"sprintIds"`
}

type JiraChangelogItem struct {
	Field      string  `json:"field"`
	From       *string `json:"from,omitempty"`
	To         *string `json:"to,omitempty"`
	FromString *string `json:"fromString,omitempty"`
	ToString   *string `json:"toString,omitempty"`
}

type JiraChangelogEvent struct {
	IssueKey  string            `json:"issueKey"`
	EventID   string            `json:"eventId"`
	Author    *JiraUser          `json:"author,omitempty"`
	CreatedAt string            `json:"createdAt"`
	Items     []JiraChangelogItem `json:"items"`
}

type JiraWorklog struct {
	IssueKey         string    `json:"issueKey"`
	WorklogID        string    `json:"worklogId"`
	Author           *JiraUser `json:"author,omitempty"`
	StartedAt        string    `json:"startedAt"`
	TimeSpentSeconds int       `json:"timeSpentSeconds"`
	CreatedAt        string    `json:"createdAt"`
	UpdatedAt        string    `json:"updatedAt"`
}

type OpsgenieTeamRef struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type CanonicalProjectWithOpsgenieTeams struct {
	Project       JiraProject       `json:"project"`
	OpsgenieTeams []OpsgenieTeamRef `json:"opsgenieTeams"`
}
